File: user-profile/src/main/java/com/oxygensened/userprofile/context/technology/TechnologyController.java
package com.oxygensened.userprofile.context.technology;

import com.oxygensened.userprofile.context.technology.dto.TechnologyRequest;
import com.oxygensened.userprofile.context.technology.dto.TechnologyView;
import io.swagger.v3.oas.annotations.tags.Tag;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.CrossOrigin;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

@Tag(name = "Technology")
@CrossOrigin
@RestController
@RequestMapping("/api/v1/users")
public class TechnologyController {
    private final TechnologyService technologyService;

    TechnologyController(TechnologyService technologyService) {
        this.technologyService = technologyService;
    }

    @PostMapping("/{userId}/technologies")
    TechnologyView add(@PathVariable Long userId,
                       @Validated @RequestBody TechnologyRequest request) {
        return technologyService.addTechnology(userId, request);
    }

    @DeleteMapping("/{userId}/technologies")
    void delete(@PathVariable Long userId,
                @Validated @RequestBody TechnologyRequest request) {
        technologyService.deleteTechnology(userId, request);
    }

}

File: static-data/src/main/kotlin/com/oxygensend/staticdata/context/technology/TechnologyController.kt
package com.oxygensend.staticdata.context.technology

import com.oxygensend.staticdata.context.PrivilegeChecker
import com.oxygensend.staticdata.context.technology.view.BaseTechnologyView
import com.oxygensend.staticdata.context.technology.view.TechnologyView
import com.oxygensend.staticdata.context.technology.view.TechnologyWithFeaturedView
import io.swagger.v3.oas.annotations.tags.Tag
import org.springframework.cache.annotation.Cacheable
import org.springframework.web.bind.annotation.*


@Tag(name = "Technology")
@RestController
@CrossOrigin
@RequestMapping("/api/v1/static-data")
internal class TechnologyController(
    private val technologyService: TechnologyService,
    private val privilegeChecker: PrivilegeChecker
) {


    @Cacheable("technologies")
    @GetMapping("/technologies")
    fun getAllTechnologies(): List<BaseTechnologyView> = technologyService.getAllTechnologies();

    @GetMapping("/technologies/details")
    fun getAllTechnologiesWithDetails(): List<TechnologyWithFeaturedView> = technologyService.getAllTechnologiesWithDetails();

    @PatchMapping("/technologies/{id}")
    fun updateTechnology(@PathVariable id: String, @RequestBody featured: Boolean): TechnologyView {
        privilegeChecker.checkEditorPrivileges()
        return technologyService.updateTechnology(id, featured)
    }

    @DeleteMapping("/technologies/{id}")
    fun deleteTechnology(@PathVariable id: String): TechnologyView {
        privilegeChecker.checkEditorPrivileges()
        return technologyService.deleteTechnology(id)
    }

    @PostMapping("/technologies")
    fun createTechnology(@RequestBody request: TechnologyRequest): TechnologyView {
        privilegeChecker.checkEditorPrivileges()
        return technologyService.createTechnology(request.name!!, request.featured!!)
    }
}